src/core/server/saved_objects/migrationsv2/actions/index.ts

```diff
@@ -419,6 +420,132 @@ export const pickupUpdatedMappings = (
     .catch(catchRetryableEsClientErrors);
 };
 
+/** @internal */
+export interface OpenPitResponse {
+  pitId: string;
+}
+
+// how long ES should keep PIT alive
+const pitKeepAlive = '10m';
+/*
+ * Creates a lightweight view of data when the request has been initiated.
+ * See https://www.elastic.co/guide/en/elasticsearch/reference/current/point-in-time-api.html
+ * */
+export const openPit = (
+  client: ElasticsearchClient,
+  index: string
+): TaskEither.TaskEither<RetryableEsClientError, OpenPitResponse> => () => {
+  return client
+    .openPointInTime({
+      index,
+      keep_alive: pitKeepAlive,
+    })
+    .then((response) => Either.right({ pitId: response.body.id }))
+    .catch(catchRetryableEsClientErrors);
+};
+
+/** @internal */
+export interface ReadWithPit {
+  outdatedDocuments: SavedObjectsRawDoc[];
+  readonly lastHitSortValue: number[] | undefined;
+}
+
+/*
+ * Requests documents from the index using PIT mechanism.
+ * Filter unusedTypesToExclude documents out to exclude them from being migrated.
+ * */
+export const readWithPit = (
+  client: ElasticsearchClient,
+  pitId: string,
+  /* When reading we use a source query to exclude saved objects types which
+   * are no longer used. These saved objects will still be kept in the outdated
+   * index for backup purposes, but won't be available in the upgraded index.
+   */
+  unusedTypesQuery: Option.Option<estypes.QueryContainer>,
+  batchSize: number,
+  searchAfter?: number[]
+): TaskEither.TaskEither<RetryableEsClientError, ReadWithPit> => () => {
+  return client
+    .search<SavedObjectsRawDoc>({
+      body: {
+        // Sort fields are required to use searchAfter, so we set some defaults here
+        sort: {
+          updated_at: { order: 'desc' },
```

Comment:
From the docs it seems like sorting with `{"_shard_doc": "desc"}` will give the best performance, but it also sounds like if you don't specify any sort this will automatically be applied.

Also sounds like for best performance we should switch off tracking total hits with `"track_total_hits": false` (maybe doesn't make a huge difference for our scale, but we're not using it anyway so might as well disable)

Reply:
> "track_total_hits": false

Actually, it might be useful to log `migrated X docs from total Y`. WDYT?

Reply:
added `"track_total_hits": false`

Reply:
we can add it in a follow-up

Reply:
++ to the follow-up. In that follow-up, do you think it's worth running an `agg` query to list the number of docs per type and trigger 1 migration _thread_ per type? Would that help with performance at all or, since Node.js is single-threaded, it would only cause more event-loop blockages?

Reply:
created https://github.com/elastic/kibana/issues/97956